File: urbanairship-core/src/main/java/com/urbanairship/actions/SetAttributesAction.java
/* Copyright Airship and Contributors */

package com.urbanairship.actions;

import com.urbanairship.Logger;
import com.urbanairship.UAirship;
import com.urbanairship.channel.AttributeEditor;
import com.urbanairship.json.JsonValue;

import java.util.Date;
import java.util.Map;

import androidx.annotation.NonNull;

/**
 * An action that sets attributes.
 * <p>
 * Accepted situations: all
 * <p>
 * Accepted argument value types: A JSON payload for setting or removing attributes. An example JSON payload:
 * <pre>
 * {
 *   "channel": {
 *     set: {"key_1": value_1, "key_2": value_2},
 *     remove: ["attribute_1", "attribute_2", "attribute_3"]
 *   },
 *   "named_user": {
 *     set: {"key_4": value_4, "key_5": value_5},
 *     remove: ["attribute_4", "attribute_5", "attribute_6"]
 *   }
 * </pre>
 */
public class SetAttributesAction extends Action {

    /**
     * Default registry name
     */
    @NonNull
    public static final String DEFAULT_REGISTRY_NAME = "set_attributes_action";

    /**
     * Default registry short name
     */
    @NonNull
    public static final String DEFAULT_REGISTRY_SHORT_NAME = "^a";

    /**
     * JSON key for channel attributes changes.
     */
    @NonNull
    private static final String CHANNEL_KEY = "channel";

    /**
     * JSON key for named user attributes changes.
     */
    @NonNull
    private static final String NAMED_USER_KEY = "named_user";

    /**
     * JSON key for setting attributes.
     */
    @NonNull
    private static final String SET_KEY = "set";

    /**
     * JSON key for removing attributes.
     */
    @NonNull
    private static final String REMOVE_KEY = "remove";

    @NonNull
    @Override
    public ActionResult perform(@NonNull ActionArguments arguments) {
        if (arguments.getValue().getMap() != null) {

            // Channel Attribute
            if (arguments.getValue().getMap().containsKey(CHANNEL_KEY)) {
                AttributeEditor channelAttributeEditor = UAirship.shared().getChannel().editAttributes();
                for (Map.Entry<String, JsonValue> entry : arguments.getValue().getMap().opt(CHANNEL_KEY).optMap().getMap().entrySet()) {
                    handleAttributeActions(channelAttributeEditor, entry);
                }
                channelAttributeEditor.apply();
            }

            // Contact Attribute
            if (arguments.getValue().getMap().containsKey(NAMED_USER_KEY)) {
                AttributeEditor contactEditor = UAirship.shared().getContact().editAttributes();
                for (Map.Entry<String, JsonValue> entry : arguments.getValue().getMap().opt(NAMED_USER_KEY).optMap().getMap().entrySet()) {
                    handleAttributeActions(contactEditor, entry);
                }
                contactEditor.apply();
            }

        }

        return ActionResult.newEmptyResult();
    }

    @Override
    public boolean acceptsArguments(@NonNull ActionArguments arguments) {
        if (arguments.getValue().isNull()) {
            return false;
        }

        if (arguments.getValue().getMap() == null) {
            return false;
        }

        // Channel attributes
        JsonValue channel = arguments.getValue().getMap().opt(CHANNEL_KEY);
        if (channel != JsonValue.NULL && !areAttributeMutationsValid(channel)) {
            return false;
        }

        // Named User attributes
        JsonValue namedUser = arguments.getValue().getMap().opt(NAMED_USER_KEY);
        if (namedUser != JsonValue.NULL && !areAttributeMutationsValid(namedUser)) {
            return false;
        }

        return channel != JsonValue.NULL || namedUser != JsonValue.NULL;
    }

    private boolean areAttributeMutationsValid(@NonNull JsonValue attributeMutations) {
        if (attributeMutations.getMap() == null) {
            return false;
        }

        JsonValue set = attributeMutations.optMap().opt(SET_KEY);
        if (set != JsonValue.NULL && !isSetAttributeMutationValid(set)) {
            return false;
        }

        JsonValue remove = attributeMutations.optMap().opt(REMOVE_KEY);
        if (remove != JsonValue.NULL && !isRemoveAttributeMutationValid(remove)) {
            return false;
        }

        return true;
    }

    private boolean isSetAttributeMutationValid(@NonNull JsonValue setAttributeMutation) {
        return setAttributeMutation.getMap() != null;
    }

    private boolean isRemoveAttributeMutationValid(@NonNull JsonValue removeAttributeMutation) {
        return removeAttributeMutation.getList() != null;
    }

    /**
     * Handles the attributes updates
     * @param attributeEditor The attribute editor
     * @param entry The attribute entry
     */
    private void handleAttributeActions(@NonNull AttributeEditor attributeEditor, @NonNull Map.Entry<String, JsonValue> entry) {
        switch (entry.getKey()) {
            case SET_KEY:
                for (Map.Entry<String, JsonValue> setAttributeEntry : entry.getValue().optMap().entrySet()) {
                    setAttribute(attributeEditor, setAttributeEntry.getKey(), setAttributeEntry.getValue().getValue());
                }
                break;
            case REMOVE_KEY:
                for (JsonValue jsonValue : entry.getValue().optList().getList()) {
                    attributeEditor.removeAttribute(jsonValue.optString());
                }
                break;
            default:
                break;
        }
    }

    /**
     * Apply the attribute settings.
     * @param attributeEditor The attribute editor.
     * @param key The attribute key.
     * @param value The attribute value.
     */
    private void setAttribute(@NonNull AttributeEditor attributeEditor, @NonNull String key, @NonNull Object value) {
        if (value instanceof Integer) {
            attributeEditor.setAttribute(key, (int) value);
        } else if (value instanceof Long) {
            attributeEditor.setAttribute(key, (long) value);
        } else if (value instanceof Float) {
            attributeEditor.setAttribute(key, (float) value);
        } else if (value instanceof Double) {
            attributeEditor.setAttribute(key, (double) value);
        } else if (value instanceof String) {
            attributeEditor.setAttribute(key, (String) value);
        } else if (value instanceof Date) {
            attributeEditor.setAttribute(key, (Date) value);
        } else {
            Logger.warn("SetAttributesAction - Invalid value type for the key: %s", key);
        }
    }

    /**
     * Default {@link SetAttributesAction.SetAttributesPredicate} predicate.
     */
    public static class SetAttributesPredicate implements ActionRegistry.Predicate {

        @Override
        public boolean apply(@NonNull ActionArguments arguments) {
            return Action.SITUATION_PUSH_RECEIVED != arguments.getSituation();
        }

    }

}

File: urbanairship-core/src/main/java/com/urbanairship/actions/SetAttributesAction.kt
/* Copyright Airship and Contributors */
package com.urbanairship.actions

import com.urbanairship.Airship
import com.urbanairship.UALog
import com.urbanairship.actions.ActionResult.Companion.newEmptyResult
import com.urbanairship.channel.AirshipChannel
import com.urbanairship.contacts.Contact
import com.urbanairship.json.JsonException
import com.urbanairship.json.JsonMap
import com.urbanairship.json.JsonSerializable
import com.urbanairship.json.JsonValue
import com.urbanairship.json.jsonMapOf
import com.urbanairship.json.requireField
import com.urbanairship.json.toJsonMap
import java.util.Date

/**
 * An action that sets attributes.
 *
 * Accepts two argument formats:
 *
 * **Dictionary format**
 * ```json
 * {
 *     "channel": {
 *         "set": { "key": "value" },
 *         "remove": ["attribute"]
 *     },
 *     "named_user": {
 *         "set": { "key": "value" },
 *         "remove": ["attribute"]
 *     }
 * }
 * ```
 *
 * **Array format** — each element describes a single operation:
 * ```json
 * [
 *     { "action": "set",    "type": "channel",    "name": "color",   "value": "blue" },
 *     { "action": "remove", "type": "channel",    "name": "color" },
 *     { "action": "set",    "type": "named_user", "name": "score",   "value": 42 },
 *     { "action": "set",    "type": "channel",    "name": "attr#id", "value": { "key": "val", "exp": 1779840000 } }
 * ]
 * ```
 * For JSON object values, `name` must be `"attributeName#instanceId"` with both parts non-empty.
 * The optional `"exp"` key inside the object is a Unix timestamp (seconds) for the expiration date.
 *
 * Accepted situations: all
 */
public class SetAttributesAction public constructor(
    private val channelProvider: () -> AirshipChannel = { Airship.channel },
    private val contactProvider: () -> Contact = { Airship.contact }
) : Action() {

    override fun perform(arguments: ActionArguments): ActionResult {
        val actions = try {
            parseActions(arguments.value.toJsonValue())
        } catch (ex: kotlin.Exception) {
            UALog.w(ex) { "Failed to parse actions" }
            return newEmptyResult()
        }

        val channelEditor = channelProvider().editAttributes()
        val contactEditor = contactProvider().editAttributes()

        val targets = actions.map { it.editor }.toSet()

        for (item in actions) {
            val editor = when (item.editor) {
                AttributeActionArgs.Editor.CHANNEL -> channelEditor
                AttributeActionArgs.Editor.CONTACT -> contactEditor
            }

            when(item) {
                is AttributeActionArgs.Remove -> editor.removeAttribute(item.name)
                is AttributeActionArgs.Set -> {
                    when(val unwrapped = item.value) {
                        is AttributeActionArgs.Value.DateValue -> editor.setAttribute(item.name, unwrapped.value)
                        is AttributeActionArgs.Value.NumberValue -> {
                            when(val number = unwrapped.value) {
                                is Int -> editor.setAttribute(item.name, number)
                                is Long -> editor.setAttribute(item.name, number)
                                is Float -> editor.setAttribute(item.name, number)
                                is Double -> editor.setAttribute(item.name, number)
                            }
                        }
                        is AttributeActionArgs.Value.StringValue -> editor.setAttribute(item.name, unwrapped.value)
                        is AttributeActionArgs.Value.Json -> editor.setAttribute(
                            attribute = unwrapped.attributeName,
                            instanceId = unwrapped.instanceId,
                            expiration = unwrapped.expiration,
                            json = unwrapped.value
                        )
                    }
                }
            }
        }

        if (targets.contains(AttributeActionArgs.Editor.CHANNEL)) {
            channelEditor.apply()
        }

        if (targets.contains(AttributeActionArgs.Editor.CONTACT)) {
            contactEditor.apply()
        }

        return newEmptyResult()
    }

    override fun acceptsArguments(arguments: ActionArguments): Boolean {
        try {
            val actions = parseActions(arguments.value.toJsonValue())
            return actions.isNotEmpty()
        } catch (_: Exception) {
            return false
        }
    }

    @Throws(JsonException::class, IllegalArgumentException::class)
    private fun parseActions(value: JsonValue): List<AttributeActionArgs> {
        if (value.isNull) {
            throw IllegalArgumentException("Null is not allowed")
        }

        if (value.isJsonList) {
            return value.requireList().list.map(AttributeActionArgs::fromJson)
        }

        val map = value.map?.map ?: throw IllegalArgumentException("json map is expected")

        val convertToAttribute: (AttributeActionArgs.Editor, JsonMap?) -> List<AttributeActionArgs> = convertToAttribute@ { editor, source ->
            if (source == null) {
                return@convertToAttribute emptyList()
            }

            val sets = source.map[SET_KEY]
                ?.requireMap()
                ?.map { (key, value) ->
                    AttributeActionArgs.Set(
                        editor = editor,
                        name = key,
                        value = AttributeActionArgs.Value.fromMapValue(value.value)
                    )
                }
                ?: emptyList()

            val removes = source.map[REMOVE_KEY]
                ?.requireList()
                ?.list
                ?.map { AttributeActionArgs.Remove(editor, it.requireString()) }
                ?: emptyList()

            sets + removes
        }

        return convertToAttribute(AttributeActionArgs.Editor.CHANNEL, map[CHANNEL_KEY]?.requireMap()) +
                convertToAttribute(AttributeActionArgs.Editor.CONTACT, map[NAMED_USER_KEY]?.requireMap())
    }

    private sealed class AttributeActionArgs(
        val action: Action,
        val editor: Editor,
        val name: String
    ): JsonSerializable {

        class Set(
            editor: Editor, name: String, val value: Value
        ) : AttributeActionArgs(Action.SET, editor, name)

        class Remove(
            editor: Editor, name: String
        ) : AttributeActionArgs(Action.REMOVE, editor, name)

        override fun toJsonValue(): JsonValue {
            val builder = JsonMap.newBuilder().put(KEY_ACTION, action).put(KEY_TARGET, editor)
                .put(KEY_NAME, name)

            when (this) {
                is Set -> builder.put(KEY_VALUE, value)
                else -> {}
            }

            return builder.build().toJsonValue()
        }

        enum class Action(val jsonValue: String) : JsonSerializable { SET("set"), REMOVE("remove");

            override fun toJsonValue(): JsonValue = JsonValue.wrap(jsonValue)

            companion object {

                @Throws(JsonException::class)
                fun fromJson(value: JsonValue): Action {
                    val content = value.requireString()
                    return entries.firstOrNull { it.jsonValue == content }
                        ?: throw JsonException("invalid value $value")
                }
            }
        }

        enum class Editor(val jsonValue: String) :
            JsonSerializable {

            CHANNEL("channel"), CONTACT("contact");

            override fun toJsonValue(): JsonValue = JsonValue.wrap(jsonValue)

            companion object {

                @Throws(JsonException::class)
                fun fromJson(value: JsonValue): Editor {
                    val content = value.requireString()
                    return entries.firstOrNull { it.jsonValue == content }
                        ?: throw JsonException("invalid value $value")
                }
            }
        }

        sealed class Value() : JsonSerializable { data class StringValue(val value: String) :
            Value()

            data class NumberValue(val value: Number) : Value()
            data class DateValue(val value: Date) : Value()
            data class Json(
                val attributeName: String,
                val instanceId: String,
                val expiration: Date?,
                val value: JsonMap
            ) : Value()

            override fun toJsonValue(): JsonValue {
                return when (this) {
                    is StringValue -> JsonValue.wrap(value)
                    is NumberValue -> JsonValue.wrap(value)
                    is DateValue -> JsonValue.wrap(value.time)
                    is Json -> {
                        val content = JsonMap.newBuilder()
                            .putAll(value)
                            .putOpt(KEY_EXPIRATION, expiration?.let { it.time / 1000 })
                            .build()
                        content.toJsonValue()
                    }
                }
            }

            companion object {

                private const val KEY_EXPIRATION = "exp"

                @Throws(JsonException::class)
                fun fromJson(value: JsonValue, name: String? = null): Value {
                    return when (val converted = value.value) {
                        is String -> StringValue(converted)
                        is Long -> DateValue(Date(converted * 1000))
                        is Number -> NumberValue(converted)
                        is JsonSerializable -> {
                            val attrName = name
                                ?: throw JsonException("Name is required for JSON object attribute values")
                            val components = attrName.split("#")
                            if (components.size != 2 || components.any { it.isEmpty() }) {
                                throw JsonException("Invalid name format: $attrName")
                            }
                            val data = converted.toJsonValue().requireMap().map.toMutableMap()
                            val expiration = data.remove(KEY_EXPIRATION)?.number
                                ?.let { Date(it.toLong() * 1000) }
                            Json(
                                attributeName = components[0],
                                instanceId = components[1],
                                expiration = expiration,
                                value = data.toJsonMap()
                            )
                        }
                        else -> throw JsonException("Unsupported value type: $converted")
                    }
                }

                @Throws(IllegalArgumentException::class)
                fun fromMapValue(value: Any?): Value {
                    return when (value) {
                        is String -> StringValue(value)
                        is Number -> NumberValue(value)
                        is Date -> DateValue(value)
                        else -> throw IllegalArgumentException("Unsupported value type: $value")
                    }
                }
            }
        }

        companion object {

            private const val KEY_ACTION = "action"
            private const val KEY_TARGET = "type"
            private const val KEY_NAME = "name"
            private const val KEY_VALUE = "value"

            @Throws(JsonException::class)
            fun fromJson(value: JsonValue): AttributeActionArgs {
                val content = value.requireMap()
                val name: String = content.requireField(KEY_NAME)
                val editor = Editor.fromJson(content.require(KEY_TARGET))

                return when (Action.fromJson(content.require(KEY_ACTION))) {
                    Action.SET -> Set(
                        editor = editor,
                        name = name,
                        value = Value.fromJson(content.require(KEY_VALUE), name)
                    )

                    Action.REMOVE -> Remove(editor, name)
                }
            }
        }
    }

    /**
     * Default [SetAttributesAction.SetAttributesPredicate] predicate.
     */
    public class SetAttributesPredicate public constructor() : ActionPredicate {

        override fun apply(arguments: ActionArguments): Boolean {
            return Situation.PUSH_RECEIVED != arguments.situation
        }
    }

    public companion object {

        /**
         * Default action names.
         */
        public val DEFAULT_NAMES: Set<String> = setOf("set_attributes_action", "modify_attributes_action", "^a")

        /**
         * JSON key for channel attributes changes.
         */
        private const val CHANNEL_KEY = "channel"

        /**
         * JSON key for named user attributes changes.
         */
        private const val NAMED_USER_KEY = "named_user"

        /**
         * JSON key for setting attributes.
         */
        private const val SET_KEY = "set"

        /**
         * JSON key for removing attributes.
         */
        private const val REMOVE_KEY = "remove"
    }
}
